runtime/src/main/java/org/corfudb/runtime/view/LayoutView.java

```diff
@@ -126,66 +124,64 @@ public Layout prepare(long epoch, long rank)
                     return cf;
                 })
                 .toArray(CompletableFuture[]::new);
-        LayoutPrepareResponse[] acceptList;
+
+        long promises = 0L;
         long timeouts = 0L;
+        long outranks = 0L;
         long wrongEpochRejected = 0L;
-        while (true) {
-            // do we still have enough for a quorum?
-            if (prepareList.length < getQuorumNumber()) {
-                log.debug("prepare: Quorum unreachable, remaining={}, required={}", prepareList,
-                        getQuorumNumber());
-                throw new QuorumUnreachableException(prepareList.length, getQuorumNumber());
+        long outrankedRecord = Long.MIN_VALUE;
+        for (CompletableFuture cf : prepareList) {
+            if (promises >= getQuorumNumber()) {
+                break;
             }
 
-            // wait for someone to complete.
             try {
-                CFUtils.getUninterruptibly(CompletableFuture.anyOf(prepareList),
-                        OutrankedException.class, TimeoutException.class, NetworkException.class,
-                        WrongEpochException.class);
+                CFUtils.getUninterruptibly(cf, OutrankedException.class, TimeoutException.class,
+                        NetworkException.class, WrongEpochException.class);
+                promises++;
             } catch (TimeoutException | NetworkException e) {
                 timeouts++;
+            } catch (OutrankedException oe) {
+                outranks++;
+                outrankedRecord = Long.max(outrankedRecord, oe.getNewRank());
             } catch (WrongEpochException we) {
                 wrongEpochRejected++;
             }
+        }
 
-            // remove errors.
-            prepareList = stream(prepareList)
-                    .filter(x -> !x.isCompletedExceptionally())
-                    .toArray(CompletableFuture[]::new);
-            // count successes.
-            acceptList = stream(prepareList)
-                    .map(x -> {
-                        try {
-                            return x.getNow(null);
-                        } catch (Exception e) {
-                            return null;
-                        }
-                    })
-                    .filter(x -> x != null)
-                    .toArray(LayoutPrepareResponse[]::new);
-
-            log.debug("prepare: Successful responses={}, needed={}, timeouts={}, "
-                            + "wrongEpochRejected={}",
-                    acceptList.length, getQuorumNumber(), timeouts, wrongEpochRejected);
-
-            if (acceptList.length >= getQuorumNumber()) {
-                break;
+        // when quorum is unreachable, check if outranks is majority and throw OutrankedException
+        if (promises < getQuorumNumber()) {
```

Comment:
I know its not your code, but it would be nice if you address this issue. 

Promises can be generated from a layout that is different from the one that is being used by `getQuorumNumber`.

Reply:
Should I note down the quorum number when sending requests? Then they are consistent.